listeners.js

```diff
@@ -0,0 +1,166 @@
+        const listeners = ui_schema["listener"] || ui_schema["logic:listener"]
+        if(listeners){
+            acc[path] = (acc[path] || [])
+            if (_.isArray(listeners)) {
+                acc[path].concat(listeners)
+            }
+            else{
+                acc[path].push(listeners)
+            }
+        }
+
+        if(ui_schema["fields"]) {
+            _.each(ui_schema["fields"], function (value, key) {
+                this._getListenersFromUISchema(value, path + '/' + key, acc)
+            }.bind(this))
+        }
+
+        else if (ui_schema["item"]) {
+            if (ui_schema["item"]["fields"]) {
+                _.each(ui_schema["item"]["fields"], function (value, key) {
+                    this._getListenersFromUISchema(value, path + '/' + key, acc)
+                }.bind(this))
+            }
+        }
+    }
+
+    // Value is the entire store
+    // Path is a json pointer to what changed
+    update(value, path) {
+        let form = Object.assign({}, value)
+        // Remove integers from path which are array element pointers and
+        // unncessary for the path listeners.
+        let jsonPointerPath =
+            _.reduce(path,
+                function (acc, piece) {
+                    return _.isString(piece) ? `${acc}/${piece}` : acc;
+                },
+                '')
+        _.each(this.listeners, (listenerInfo, key) => {
+            // We want to check for the current key, but also any parents that are listening
+            if (_.startsWith(jsonPointerPath, key)) {
+                const formSubSection = jsonpointer.get(form, key)
+                const {schema, ui_schema} = this._getSchemaAtPath(key)
+                _.each(listenerInfo, (val) => this._fireListenerLogicAndSetState(val, formSubSection, schema, ui_schema))
+            }
+        })
+    }
+
+    updateAll(value) {
+        let form = Object.assign({}, value)
+        _.each(this.listeners, (listenerInfo, key) => {
+            const formSubSection = jsonpointer.get(form, key)
+            const {schema, ui_schema} = this._getSchemaAtPath(key)
+            _.each(listenerInfo, (val) => this._fireListenerLogicAndSetState(val, formSubSection, schema, ui_schema))
+        })
+    }
+
+    _fireListenerLogicAndSetState = (val, storeSubSection, schema, ui_schema) => {
+        let output = null
+        if(val["function"] == "list_adder") {
+            output = this.listNumberAdder(storeSubSection, schema, ui_schema)
+        } else if(val["function"] == "sub_list_adder") {
+            const subListPath = val["sub_list_path"]
+            output = this.subListNumberAdder(storeSubSection, subListPath)
+        } else if(val["function"] == "eval") {
+            output = this.evalForm(storeSubSection, val["eval"])
+        }
+        this._maybeUpdateComponentStateInsideForm(val.output_key, output)
+    }
+
+    _maybeUpdateComponentStateInsideForm = (outputKey, output) => {
```

Comment:
"maybe" is a code smell to me-- `updateIfNeeded` would be a more trustworthy name but each suggests information leakage: your callers shouldn't care about the implementation or when things need to get updated, they can blindly call an `updateComponentStateInsideForm` and let it worry about whether things happen or not.